File: mutual_friends.py
from pyspark import SparkContext

def parse_line_safe(line):
    try:
        if line.startswith("#") or not line.strip():
            return None
        parts = line.strip().split("\t")
        if len(parts) != 3:
            return None
        user_id = parts[0].strip()
        name = parts[1].strip()
        friends = parts[2].split(",")
        return (user_id, name, [f.strip() for f in friends if f.strip()])
    except Exception:
        return None

# 🔁 Création symétrique des paires (pour chaque ami, on émet la paire)
def make_symmetric_pairs(record):
    user_id, _, friends = record
    result = []
    for friend in friends:
        key = tuple(sorted([user_id, friend]))  # (min, max)
        result.append((key, set(friends)))
    return result

def main():
    sc = SparkContext(appName="MutualFriends")
    
    lines = sc.textFile("file:///home/daouda/tp2/data.txt")
    
    parsed = lines.map(parse_line_safe).filter(lambda x: x is not None)
    
    id_to_name = parsed.map(lambda x: (x[0], x[1])).collectAsMap()
    
    # Générer les paires symétriques
    pairs = parsed.flatMap(make_symmetric_pairs)

    # Faire intersection
    mutual = pairs.reduceByKey(lambda x, y: x & y)

    target = tuple(sorted(["1", "2"]))  # Mohamed & Sidi
    result = mutual.filter(lambda x: x[0] == target).collect()

    with open("test_output.txt", "w") as f:
        if result and result[0][1]:
            names = [id_to_name.get(fid, fid) for fid in sorted(result[0][1])]
            f.write(f"Ami(s) commun(s) entre {target[0]}<{id_to_name[target[0]]}> et {target[1]}<{id_to_name[target[1]]}> : {', '.join(names)}\n")
        else:
            f.write("Aucun ami commun trouvé.\n")

    sc.stop()

if __name__ == "__main__":
    main()

File: MutualFriends.scala
// Fonction pour générer les couples triés
def pairs(str: Array[String]) = {
  val user = str(0)
  val friends = str(1).split(",")
  for (friend <- friends) yield {
    val pair = if (user.toInt < friend.toInt) (user, friend) else (friend, user)
    (pair, friends)
  }
}

// Charger les données depuis sample.txt
val data = sc.textFile("sample.txt")
val data1 = data.map(_.split("\t")).filter(x => x.length == 2)
val pairCounts = data1.flatMap(pairs).reduceByKey((a, b) => a.intersect(b))

// Formater le résultat
val p1 = pairCounts.map {
  case ((u1, u2), common) => s"$u1\t$u2\t${common.mkString(",")}"
}

// Sauvegarder résultat global
p1.saveAsTextFile("output")

// Extraire les résultats spécifiques
var ans = ""
val queries = Seq(("0", "4"), ("20", "22939"), ("1", "29826"), ("19272", "6222"), ("28041", "28056"))

for ((u1, u2) <- queries) {
  val result = p1.map(_.split("\t"))
    .filter(x => x.length == 3 && x(0) == u1 && x(1) == u2)
    .flatMap(x => x(2).split(",")).collect()

  ans += s"$u1\t$u2\t${result.mkString(",")}\n"
}

val answer = sc.parallelize(Seq(ans))
answer.saveAsTextFile("output1")
